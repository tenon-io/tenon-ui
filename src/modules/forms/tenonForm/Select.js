import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

/**
 * @component
 * A default view component to use with the tenon-ui
 * smart select controller.
 *
 * @prop {string} contentHintText - Text to show as content
 *                hint of the select.
 * @prop {string} errorText - Text to show as error message
 *                of the select.
 * @prop required {function} getLabelProps - Prop getter function
 *              from the smart select controller for the <label>.
 * @prop required {function} getSelectProps - Prop getter function
 *              from the smart select controller for the <select>.
 * @prop {function} getErrorProps - Prop getter function
 *              from the smart select controller for the
 *              error message container.
 * @prop {function} getContentHintProps - Prop getter
 *              function from the smart select controller
 *              for the content hint container.
 * @prop required {string} labelText
 * @prop {object} labelProps - Control prop collection
 *              for the <label>
 * @prop {boolean} showError - A flag to indicate when the
 *              error container should be displayed.
 * @prop {string} className - An optional class string to transfer to the
 *              className prop of the select element.
 * @prop required {React element} children - The <option> tags to render for the
 *              <select> element.
 *
 * NOTE: All props given to this select that does not
 * satisfy one of the above will be passed into the the
 * prop getter function of the input to enrich the select or
 * override already configured default props. Providing
 * extraneous properties that should not be rendered on an
 * <select> in the DOM can create run time errors.
 */
const Select = ({
    contentHintText,
    errorText,
    getLabelProps,
    getSelectProps,
    getErrorProps,
    getContentHintProps,
    labelText,
    labelProps,
    showError,
    className,
    children,
    ...rest
}) => (
    <div className="form-group">
        <div className="field-wrapper">
            <label {...getLabelProps(labelProps)}>{labelText}</label>
            <select
                className={
                    classNames(className, { 'has-error': showError }) || null
                }
                {...getSelectProps(rest)}
            >
                {children}
            </select>
        </div>
        {contentHintText && getContentHintProps ? (
            <div className="info-wrapper">
                <span {...getContentHintProps()}>{contentHintText}</span>
            </div>
        ) : null}
        {showError && getErrorProps ? (
            <div className="error-wrapper">
                <span {...getErrorProps()}>{errorText}</span>
            </div>
        ) : null}
    </div>
);

Select.displayName = 'Select';

Select.propTypes = {
    children: PropTypes.node.isRequired,
    contentHintText: PropTypes.string,
    errorText: PropTypes.string,
    getLabelProps: PropTypes.func.isRequired,
    getSelectProps: PropTypes.func.isRequired,
    getErrorProps: PropTypes.func,
    getContentHintProps: PropTypes.func,
    labelText: PropTypes.string.isRequired,
    labelProps: PropTypes.object,
    showError: PropTypes.bool,
    className: PropTypes.string
};

export default Select;
