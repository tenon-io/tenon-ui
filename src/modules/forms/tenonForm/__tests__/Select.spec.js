import React, { createRef } from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import Select from '../Select';

describe('Select', () => {
    afterEach(cleanup);

    it('should render an input and label only', () => {
        const { container } = render(
            <Select
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getSelectProps={() => ({ id: 'foo' })}
                labelText="Test select"
            >
                <option value="one">One</option>
                <option value="two">Two</option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render an error text when invalid', () => {
        const { container } = render(
            <Select
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getSelectProps={() => ({
                    id: 'foo',
                    'aria-describedby': 'bar'
                })}
                getErrorProps={() => ({ id: 'bar' })}
                labelText="Test select"
                errorText="Test error"
                showError={true}
            >
                <option value="one">One</option>
                <option value="two">Two</option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a content hint if given', () => {
        const { container } = render(
            <Select
                contentHintText="Content hint test"
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getSelectProps={() => ({
                    id: 'foo',
                    'aria-describedby': 'bar'
                })}
                getContentHintProps={() => ({ id: 'bar' })}
                labelText="Test select"
            >
                <option value="one">One</option>
                <option value="two">Two</option>
            </Select>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should allow for easy overriding of default props', () => {
        const { getByLabelText, getByText } = render(
            <Select
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getSelectProps={() => ({ id: 'foo' })}
                labelText="Test select"
                id="foo"
                labelProps={{ htmlFor: 'foo' }}
            >
                {' '}
                <option value="one">One</option>
                <option value="two">Two</option>
            </Select>
        );

        expect(getByLabelText('Test select')).toHaveAttribute('id', 'foo');
        expect(getByText('Test select')).toHaveAttribute('for', 'foo');
    });

    it('should allow easy focus of the internal select', () => {
        const selectRef = createRef();

        const { getByLabelText } = render(
            <Select
                ref={selectRef}
                getLabelProps={() => ({ htmlFor: 'foo' })}
                getSelectProps={() => ({ id: 'foo' })}
                labelText="Test select"
            >
                <option value="one">One</option>
                <option value="two">Two</option>
            </Select>
        );

        expect(getByLabelText('Test select')).toEqual(selectRef.current);
    });
});
