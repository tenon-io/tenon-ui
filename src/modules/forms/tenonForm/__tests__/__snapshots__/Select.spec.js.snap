// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render a content hint if given 1`] = `
<div
  class="form-group"
>
  <div
    class="field-wrapper"
  >
    <label
      for="foo"
    >
      Test select
    </label>
    <select
      aria-describedby="bar"
      id="foo"
    >
      <option
        value="one"
      >
        One
      </option>
      <option
        value="two"
      >
        Two
      </option>
    </select>
  </div>
  <div
    class="info-wrapper"
  >
    <span
      id="bar"
    >
      Content hint test
    </span>
  </div>
</div>
`;

exports[`Input should render an error text when invalid 1`] = `
<div
  class="form-group"
>
  <div
    class="field-wrapper"
  >
    <label
      for="foo"
    >
      Test select
    </label>
    <select
      aria-describedby="bar"
      class="has-error"
      id="foo"
    >
      <option
        value="one"
      >
        One
      </option>
      <option
        value="two"
      >
        Two
      </option>
    </select>
  </div>
  <div
    class="error-wrapper"
  >
    <span
      id="bar"
    >
      Test error
    </span>
  </div>
</div>
`;

exports[`Input should render an input and label only 1`] = `
<div
  class="form-group"
>
  <div
    class="field-wrapper"
  >
    <label
      for="foo"
    >
      Test select
    </label>
    <select
      id="foo"
    >
      <option
        value="one"
      >
        One
      </option>
      <option
        value="two"
      >
        Two
      </option>
    </select>
  </div>
</div>
`;
